src/ReplicationManager.ts

```diff
@@ -0,0 +1,150 @@
+import NetworkPeer from './NetworkPeer'
+import HypercoreProtocol from 'hypercore-protocol'
+import FeedStore, { FeedId } from './FeedStore'
+import PeerConnection from './PeerConnection'
+import { getOrCreate, encodeDiscoveryId, DiscoveryId, toDiscoveryId, joinSets } from './Misc'
+import MessageCenter, { Routed } from './MessageCenter'
+import pump from 'pump'
+import MapSet from './MapSet'
+import Queue from './Queue'
+
+type ReplicationMsg = DiscoveryIdsMsg
+
+interface DiscoveryIdsMsg {
+  type: 'DiscoveryIds'
+  discoveryIds: DiscoveryId[]
+}
+
+export interface Discovery {
+  feedId: FeedId
+  discoveryId: DiscoveryId
+  peer: NetworkPeer
+}
+
+export default class ReplicationManager {
+  private protocols: WeakMap<PeerConnection, HypercoreProtocol>
+  private feeds: FeedStore
+
+  discoveryIds: Map<DiscoveryId, FeedId>
+  messages: MessageCenter<ReplicationMsg>
+  peers: Set<NetworkPeer>
+  peersByDiscoveryId: MapSet<DiscoveryId, NetworkPeer>
+
+  discoveryQ: Queue<Discovery>
+
+  constructor(feeds: FeedStore) {
+    this.discoveryIds = new Map()
+    this.protocols = new WeakMap()
+    this.peers = new Set()
+    this.peersByDiscoveryId = new MapSet()
+    this.discoveryQ = new Queue('ReplicationManager:discoveryQ')
+    this.feeds = feeds
+    this.messages = new MessageCenter('ReplicationManager')
+    this.messages.inboxQ.subscribe(this.onMessage)
+  }
+
+  addFeedIds(feedIds: FeedId[]): void {
+    const discoveryIds: DiscoveryId[] = []
+
+    for (const feedId of feedIds) {
+      const discoveryId = toDiscoveryId(feedId)
+      if (!this.discoveryIds.has(discoveryId)) {
+        this.discoveryIds.set(discoveryId, feedId)
+        discoveryIds.push(discoveryId)
+      }
+    }
+
+    if (discoveryIds.length > 0) {
+      this.messages.sendToPeers(this.peers, {
+        type: 'DiscoveryIds',
+        discoveryIds,
+      })
+    }
+  }
+
+  getFeedId(discoveryId: DiscoveryId): FeedId | undefined {
+    return this.discoveryIds.get(discoveryId)
+  }
+
+  getPeersWith(discoveryIds: DiscoveryId[]): Set<NetworkPeer> {
+    return joinSets(discoveryIds.map((id) => this.peersByDiscoveryId.get(id)))
+  }
+
+  close(): void {
+    this.messages.inboxQ.unsubscribe()
+  }
+
+  /**
+   * Should be called when a peer connects.
```

Comment:
Is there a possibility of this not being called?

Reply:
Nah, just poor phrasing